# Exercise 1 

In this exercise, we will work with the `spam.csv` data file. This file contains text and a label attached to each text. The goal is to clean the text data and then build classifcation models. 

### Exercise 1(a) (2 points)

Load the `pandas` and `numpy` libraries.

In [1]:
import pandas as pd
import numpy as np

### Exercise 1(b) (2 points)

Read the `csv` file and create a data-frame called `spam_df`.

In [2]:
spam_df = pd.read_csv("spam.csv")
spam_df.head()

,label,text
0,ham,"Go until jurong point, crazy.. Available only ..."
1,ham,Ok lar... Joking wif u oni...
2,spam,Free entry in 2 a wkly comp to win FA Cup fina...
3,ham,U dun say so early hor... U c already then say...
4,ham,"Nah I don't think he goes to usf, he lives aro..."


### Exercise 1(c) (2 points)

Report the frequency table of `label`.

In [3]:
spam_df['label'].value_counts()
spam_df['label'].value_counts(normalize=True)

label
ham     0.865937
spam    0.134063
Name: proportion, dtype: float64

In [4]:
round(100*spam_df['label'].value_counts(normalize=True), 2)

label
ham     86.59
spam    13.41
Name: proportion, dtype: float64

### Exercise 1(d) (2 points)

Using the `string` library, remove all the punctuation of the `text` column.

In [5]:
import string

spam_df['text_clean'] = spam_df['text'].str.replace(f"[{string.punctuation}]", "", regex=True)
spam_df.head()

,label,text,text_clean
0,ham,"Go until jurong point, crazy.. Available only ...",Go until jurong point crazy Available only in ...
1,ham,Ok lar... Joking wif u oni...,Ok lar Joking wif u oni
2,spam,Free entry in 2 a wkly comp to win FA Cup fina...,Free entry in 2 a wkly comp to win FA Cup fina...
3,ham,U dun say so early hor... U c already then say...,U dun say so early hor U c already then say
4,ham,"Nah I don't think he goes to usf, he lives aro...",Nah I dont think he goes to usf he lives aroun...


### Exercise 1(e) (10 points)

Using the `nltk` library, remove stopwords and pronouns from `text_clean` and store in another column called `'text_clean_1'`. 

In [6]:
import nltk
nltk.download('punkt')
nltk.download('stopwords')

from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize

stop_words = set(stopwords.words('english'))
pronouns = ['i', 'me', 'my', 'mine', 'you', 'your', 'yours', 'he', 'him', 'his', 'she', 'her', 'hers', 'we', 'us', 'our', 'ours', 'they', 'them', 'their', 'theirs', 'u', 'ur', 'im', 'dont', '2', 'ill', '4']

def clean_text(text):
    #tokenize the text
    tokens = word_tokenize(text.lower())

    #remove unwanted tokens (stop words and pronouns)
    tokens = [token for token in tokens if token not in stop_words and token not in pronouns]

    #replacing words
    tokens = [word.replace('txt', 'text') for word in tokens]

    return " ".join(tokens)

spam_df['text_clean_1'] = spam_df['text_clean'].apply(clean_text)
spam_df.head()


[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\isabe_8m\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\isabe_8m\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


,label,text,text_clean,text_clean_1
0,ham,"Go until jurong point, crazy.. Available only ...",Go until jurong point crazy Available only in ...,go jurong point crazy available bugis n great ...
1,ham,Ok lar... Joking wif u oni...,Ok lar Joking wif u oni,ok lar joking wif oni
2,spam,Free entry in 2 a wkly comp to win FA Cup fina...,Free entry in 2 a wkly comp to win FA Cup fina...,free entry wkly comp win fa cup final tkts 21s...
3,ham,U dun say so early hor... U c already then say...,U dun say so early hor U c already then say,dun say early hor c already say
4,ham,"Nah I don't think he goes to usf, he lives aro...",Nah I dont think he goes to usf he lives aroun...,nah think goes usf lives around though


### Exercise 1(f) (10 points)

Perform Lemmatization of `text_clean_1` and store the results in another column called `text_clean_2` in the `spam_df` data frame.

In [7]:
from nltk.corpus import wordnet
from nltk.tokenize import word_tokenize
from nltk.stem import WordNetLemmatizer

#download necessary resources
nltk.download('wordnet')
nltk.download('averaged_perceptron_tagger')

#initialize lemmatizer
lemmatizer = WordNetLemmatizer()

#function to get the part of speech tag for lemmatization
def get_wordnet_pos(word):
    tag = nltk.pos_tag([word])[0][1][0].upper()
    tag_dict = {"J": wordnet.ADJ,
                "N": wordnet.NOUN,
                "V": wordnet.VERB,
                "R": wordnet.ADV}
    return tag_dict.get(tag, wordnet.NOUN)

#function to lemmatize text
def lemmatize_text(text):
    tokens = word_tokenize(text)
    lemmatized_words = [lemmatizer.lemmatize(token, get_wordnet_pos(token)) for token in tokens]
    return " ".join(lemmatized_words)

#apply the function to lemmatize the text
spam_df['text_clean_2'] = spam_df['text_clean_1'].apply(lemmatize_text)
spam_df.head()


[nltk_data] Downloading package wordnet to
[nltk_data]     C:\Users\isabe_8m\AppData\Roaming\nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     C:\Users\isabe_8m\AppData\Roaming\nltk_data...
[nltk_data]   Package averaged_perceptron_tagger is already up-to-
[nltk_data]       date!


,label,text,text_clean,text_clean_1,text_clean_2
0,ham,"Go until jurong point, crazy.. Available only ...",Go until jurong point crazy Available only in ...,go jurong point crazy available bugis n great ...,go jurong point crazy available bugis n great ...
1,ham,Ok lar... Joking wif u oni...,Ok lar Joking wif u oni,ok lar joking wif oni,ok lar joking wif oni
2,spam,Free entry in 2 a wkly comp to win FA Cup fina...,Free entry in 2 a wkly comp to win FA Cup fina...,free entry wkly comp win fa cup final tkts 21s...,free entry wkly comp win fa cup final tkts 21s...
3,ham,U dun say so early hor... U c already then say...,U dun say so early hor U c already then say,dun say early hor c already say,dun say early hor c already say
4,ham,"Nah I don't think he goes to usf, he lives aro...",Nah I dont think he goes to usf he lives aroun...,nah think goes usf lives around though,nah think go usf life around though


### Exercise 1(g) (5 points)

Based on the results from part 1(f), we see digits in some of the text in `text_clean_2`. Remove all the digits from `text_clean_2` and store the results in another column `text_clean_3` in the `spam_df` data frame.

In [8]:
import re

#function to remove numbers from text
def remove_numbers(text):
    return re.sub(r'\d+', '', text)

#apply the function to 'text_clean_2' to remove numbers from the text
spam_df['text_clean_3'] = spam_df['text_clean_2'].apply(remove_numbers)
spam_df.head()

,label,text,text_clean,text_clean_1,text_clean_2,text_clean_3
0,ham,"Go until jurong point, crazy.. Available only ...",Go until jurong point crazy Available only in ...,go jurong point crazy available bugis n great ...,go jurong point crazy available bugis n great ...,go jurong point crazy available bugis n great ...
1,ham,Ok lar... Joking wif u oni...,Ok lar Joking wif u oni,ok lar joking wif oni,ok lar joking wif oni,ok lar joking wif oni
2,spam,Free entry in 2 a wkly comp to win FA Cup fina...,Free entry in 2 a wkly comp to win FA Cup fina...,free entry wkly comp win fa cup final tkts 21s...,free entry wkly comp win fa cup final tkts 21s...,free entry wkly comp win fa cup final tkts st ...
3,ham,U dun say so early hor... U c already then say...,U dun say so early hor U c already then say,dun say early hor c already say,dun say early hor c already say,dun say early hor c already say
4,ham,"Nah I don't think he goes to usf, he lives aro...",Nah I dont think he goes to usf he lives aroun...,nah think goes usf lives around though,nah think go usf life around though,nah think go usf life around though


### Exercise 1(h) (5 points)

Using the `TfidfVectorizer` function from `sklearn.feature_extraction.text`, convert the `text_clean_3` column into a matrix of TF-IDF features. Use `max_features=30` to create the input features, and map `label` to 0-1.

In [9]:
from sklearn.feature_extraction.text import TfidfVectorizer

tdif_vectorizer = TfidfVectorizer(max_features=30)
X_tfidf = tdif_vectorizer.fit_transform(spam_df['text_clean_3']).toarray()
y = spam_df['label'].map({'ham': 0, 'spam': 1})

In [10]:
X_tfidf

array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]], shape=(5572, 30))

### Exercise 1(i) (7 points)

Using 5-fold cross-validation strategy (use `StratifiedKFold`), train and evaluate the out-of-sample performance of the `LogisticRegression` model. Use a probability threshold of 15% (i.e., classify as spam if the predicted probability is greater than or equal to 0.15) and use `recall_score` as measure of performance.

In [11]:
from sklearn.metrics import recall_score
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import StratifiedKFold
from sklearn.naive_bayes import MultinomialNB

skf = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)
threshold = 0.15
scores = []

for i, (train_index, test_index) in enumerate(skf.split(X_tfidf, y)):

    #split the data into training and testing sets
    X_train, X_test = X_tfidf[train_index], X_tfidf[test_index]
    y_train, y_test = y[train_index], y[test_index]

    #initialize and train the logistic regression model
    model = LogisticRegression(random_state=42)
    model.fit(X_train, y_train)

    #predict probabilities for the test set
    y_pred_proba = model.predict_proba(X_test)[:, 1]
    y_pred = (y_pred_proba >= threshold).astype(int)
    score = recall_score(y_test, y_pred)
    scores.append(score)
    print(f"Fold {i+1} recall score:", score)

print("Average recall score:", sum(scores)/len(scores))

Fold 1 recall score: 0.8466666666666667
Fold 2 recall score: 0.8733333333333333
Fold 3 recall score: 0.8187919463087249
Fold 4 recall score: 0.785234899328859
Fold 5 recall score: 0.8389261744966443
Average recall score: 0.8325906040268457


### Exercise 1(j) (7 points)

Using the same cross-validation strategy from part 1(i), train and evaluate the out-of-fold performance of the `MultinomialNB` model. Use a probability threshold of 15% (i.e., classify as spam if the predicted probability is greater than or equal to 0.15) and use `recall_score` as measure of performance.

In [12]:
from sklearn.naive_bayes import MultinomialNB
threshold = 0.15
scores = []

for i, (train_index, test_index) in enumerate(skf.split(X_tfidf, y)):

    #split the data into training and testing sets
    X_train, X_test = X_tfidf[train_index], X_tfidf[test_index]
    y_train, y_test = y[train_index], y[test_index]

    #initialize and train the multinomial naive bayes model
    model = MultinomialNB()
    model.fit(X_train, y_train)

    #predict probabilities for the test set
    y_pred_proba = model.predict_proba(X_test)[:, 1]
    y_pred = (y_pred_proba >= threshold).astype(int)
    score = recall_score(y_test, y_pred)
    scores.append(score)
    print(f"Fold {i+1} recall score:", score)

print("Average recall score:", sum(scores)/len(scores))

Fold 1 recall score: 0.8266666666666667
Fold 2 recall score: 0.8266666666666667
Fold 3 recall score: 0.7718120805369127
Fold 4 recall score: 0.7583892617449665
Fold 5 recall score: 0.8120805369127517
Average recall score: 0.7991230425055929


### Exercise 1(k) (8 points)

Using the same cross-validation strategy from part 1(i), train and evaluate the out-of-fold performance of a deep learning model (e.g., using `tensorflow`/`keras` or `pytorch`). Use a probability threshold of 15% (i.e., classify as spam if the predicted probability is greater than or equal to 0.15) and use `recall_score` as measure of performance.

In [14]:
!pip install torch

  Using cached torch-2.14.1-cp311-cp311-win_amd64.whl.metadata (38 kB)
Using cached torch-2.14.1-cp311-cp311-win_amd64.whl (124.1 MB)


ERROR: Could not install packages due to an OSError: [Errno 2] No such file or directory: 'C:\\Users\\isabe_8m\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python311\\site-packages\\torch\\include\\ATen\\native\\transformers\\cuda\\mem_eff_attention\\iterators\\predicated_tile_access_iterator_residual_last.h'
HINT: This error might have occurred since this system does not have Windows Long Path support enabled. You can find information on how to enable this at https://pip.pypa.io/warnings/enable-long-paths


[notice] A new release of pip is available: 24.0 -> 26.2.1
[notice] To update, run: C:\Users\isabe_8m\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip


In [16]:
import torch
import torch.nn as nn

torch.manual_seed(42)

class Spam_Net(nn.Module):
    def __init__(self, n_features):
        super().__init__()
        self.net = nn.Sequential(
            nn.Linear(n_features, 16),
            nn.ReLU(),
            nn.Linear(16, 8),
            nn.ReLU(),
            nn.Linear(8, 1),
            nn.Sigmoid()
        )

    def forward(self, x):
        return x


threshold = 0.15
scores = []

for i, (train_index, test_index) in enumerate(skf.split(X_tfidf, y)):

    #split the data into training and testing sets
    X_train, X_test = X_tfidf[train_index], X_tfidf[test_index]
    y_train, y_test = y.iloc[train_index], y.iloc[test_index]

    #convert the data to torch tensors
    X_train_t = torch.tensor(X_train, dtype=torch.float32)
    y_train_t = torch.tensor(y_train, dtype=torch.float32)
    X_test_t = torch.tensor(X_test, dtype=torch.float32)

    #define the model
    model = Spam_Net(X_train.shape[1])
    criterion = nn.BCELoss()
    optimizer = torch.optim.Adam(model.parameters(), lr=0.01)

    #train the model
    model.train()
    for epoch in range(100):
        optimizer.zero_grad()
        output = model(X_train_t)
        loss = criterion(output, y_train_t)
        loss.backward()
        optimizer.step()

    model.eval()
    with torch.no_grad():
        y_pred = model(X_test_t).numpy().ravel()
    y_pred_class = (y_pred >= threshold).astype(float)
    score = recall_score(y_test, y_pred_class)
    scores.append(score)
    print(f"Fold {i+1} recall score: {score}")

print("Average recall score:", sum(scores)/len(scores))

ImportError: DLL load failed while importing _C: The specified module could not be found.

### Exercise 1(l) (3 points)

Based on the results from parts 1(i), 1(j) and 1(k), what model would you use to predict spam label?

Based on my results, the 'LogisticRegression' and the deep learning model have essential the same out-of-fold recall (approx .83). Because 'LogisticRegression' is simpler, I would use it.